### ISSUE TITLE
Cannot input some characters with XIM

### ISSUE THREAD

**Issue by ncfavier** [ibus/ibus#2422]
## TL;DR

If you can't input `≤ ≥ √ ≈ ∙` in old X applications, a workaround is to set `LC_CTYPE=ja_JP.UTF-8` (or another CJK locale), globally or just for ibus, OR you can use this patch: https://github.com/ibus/ibus/pull/2425

-------

#### Which distribution and version?: 
NixOS 22.05

#### Which desktop environment and version?: 
No DE, just bspwm.

#### Which session type?: 
X11 (but `XDG_SESSION_TYPE=tty` because I don't use a display manager)

#### Which application and version?: 
alacritty, xterm, urxvt, NOT kitty, NOT firefox, NOT thunar. *Basically only the applications in which the ibus panel appears when I hit compose.*

#### IBus version?: 
1.5.26

#### Issue description:

There are some Unicode characters that I cannot input using IBus' XCompose layer. This is independent of the key sequence mapped to the character and only seems to happen in applications that use the panel (see above).

Some characters I **cannot** input are: `∙` (U+2219), `√` (U+221A), `≈` (U+2248), `≤` (U+2264), `≥` (U+2265).

Some characters from neighbouring ranges I **can** input are: `∛`, `∘`, `∗`, `∞`, `∀`, `∧`, `∨`, `∩`, `∪`, `∫`, `≃`, `≅`, `≡`, `≪`, `≫`.

The same issue with `√` seems to have been reported a few years ago, but not resolved: https://github.com/ibus/ibus/issues/1852#issuecomment-225424968 (cc @thedoctar).

#### Steps to reproduce:
With the default XCompose and IBus running, type `<Compose>v/` into xterm. This should fail to input anything, while in other applications it should output `√`.

#### Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
Yes.

#### Do you see any errors when you run ibus-daemon with the verbose option?:
No.

#### Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 
Haven't tried yet.

**Comment by thedoctar** [comment]
I no longer have this problem on Lubuntu 20.04.4, but it uses fcitx now.

**Comment by mike-fabian** [comment]
I can reproduce that in xterm and urxvt. It seems to happen in all clients using XIM.

**Comment by mike-fabian** [comment]
This video shows how √ (U+221A) “vanishes” from commits when typing with ibus in clients using XIM (urxvt in this video).

No matter whether the compose support in ibus or the compose support in ibus-typing-booster is used, √ (U+221A) vanishes on commit. Even when committing something like “square√root” using ibus-typing-booster, the √ seems to be “filtered out” and only “squareroot” is commited in urxvt. But it works fine in gedit which uses the Gtk3 input module.

It is also not a font problem as this can be pasted just fine into urxvt. 


https://user-images.githubusercontent.com/2330175/178115232-e8d65b31-7ef5-49d4-806f-28d1d05906db.mp4

**Comment by mike-fabian** [comment]
So it is not a problem with the compose support, https://github.com/ibus/ibus/issues/1852 had the same problem with ibus table. 

It looks like that no matter which ibus engine is used as soon as something containing √ (U+221A) is commited while XIM is used, the character √ (U+221A)  is missing in the commit.

**Comment by rohieb** [comment]
I also noticed that I cannot input the other characters (`≤≥≈√∙`) in xterm and urxvt through any other input method, e.g. the Unicode code point selector (Ctrl-Shift-u), or emoji search (Ctrl-Shift-e) with the `xkb` engine; or through the `table:latex` engine. The same works in Firefox and Inkscape though (which are GTK apps). But not in JOSM, which is a Java application (with GTK+ look and feel, I guess through Swing, but I'm not an expert…)

I'm using Debian stable, awesome-wm desktop environment, IBus 1.5.23 started via `~/.xinitrc` as `ibus-daemon --xim --daemonize`.
Environment variables:
```
XMODIFIERS=@im=ibus
GTK_IM_MODULE=ibus
QT_IM_MODULE=ibus
```
and locale output:
```
LANG=en_US.UTF-8
LANGUAGE=en_US:en
LC_CTYPE=de_DE.UTF-8
LC_NUMERIC=en_US.UTF-8
LC_TIME=de_DE.UTF-8
LC_COLLATE=de_DE.UTF-8
LC_MONETARY=de_DE.UTF-8
LC_MESSAGES=en_US.UTF-8
LC_PAPER=de_DE.UTF-8
LC_NAME=de_DE.UTF-8
LC_ADDRESS=de_DE.UTF-8
LC_TELEPHONE=de_DE.UTF-8
LC_MEASUREMENT=de_DE.UTF-8
LC_IDENTIFICATION=de_DE.UTF-8
LC_ALL=
```

**Comment by mike-fabian** [comment]
I.e. the √ is lost after trying to convert `hello√world` from UTF-8 to COMPOUND_TEXT  with Xutf8TextListToTextProperty().

**Comment by mike-fabian** [comment]
If I do the same using ∞ U+221E instead` of √ U+221A it works, i.e. when I type `hello∞world` and check with gdb, I see:

![Screenshot](https://user-images.githubusercontent.com/2330175/180175084-9d700151-9f26-4728-9554-868f5161a853.png)

**Comment by ncfavier** [comment]
Thanks for digging into this! Could we solve this by simply replacing `XCompoundTextStyle` with `XUTF8StringStyle`? (Maybe conditionally on `X_HAVE_UTF8_STRING`)

**Comment by ncfavier** [comment]
Nope, that somehow outputs latin-1 instead (é becomes Ã©)

**Comment by mike-fabian** [comment]
> Thanks for digging into this! Could we solve this by simply replacing `XCompoundTextStyle` with `XUTF8StringStyle`? (Maybe conditionally on `X_HAVE_UTF8_STRING`)

Fujiwara San says `XUTF8StringStyle` is unfortunately not supported in XIM server.

**Comment by mike-fabian** [comment]
I also found that everything works fine when a CJK locale like ja_JP.UTF-8, ko_KR.UTF-8, ... is used but it does not work in en_US.UTF-8, de_DE.UTF-8, ...

**Comment by ncfavier** [comment]
> I also found that everything works fine when a CJK locale like ja_JP.UTF-8, ko_KR.UTF-8, ... is used but it does not work in en_US.UTF-8, de_DE.UTF-8, ...

That seems to work around it indeed.

**Comment by mike-fabian** [comment]
Fujiwara San wrote this small test program `compound.c` to check Xutf8TextListToTextProperty():

```C
#include <locale.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <stdio.h>

typedef int (*Func) (Display *,
                     char **,
                     int,
                     XICCEncodingStyle,
                     XTextProperty *);

static const struct {
    char *label;
    Func  func;
} data[] = {
    { "Xutf8TextListToTextProperty", Xutf8TextListToTextProperty },
    { "XmbTextListToTextProperty", XmbTextListToTextProperty },
    { NULL, 0 }
};

int
main (int argc, char *argv[])
{
    Display *display;
    const char *text = "√";
    XTextProperty tp;
    int ret;
    int i, j;

    setlocale(LC_ALL, "");
    display = XOpenDisplay (NULL);
    for (i = 0; data[i].label; i++) {
        tp.value = 0;
        ret = data[i].func (display,
                            (char **)&(text), 1,
                            XCompoundTextStyle,
                            &tp);
        if (tp.value) {
            printf ("%s ret:%d value: ", data[i].label, ret);
            for (j = 0; tp.value[j]; j++)
                printf ("%x ", tp.value[j]);
            printf ("\n");
        } else {
            printf ("%s ret:%d value:0\n", data[i].label, ret);
        }
    }
    return 0;
}
```

**Comment by mike-fabian** [comment]
The test program also shows that it works in ja_JP.UTF-8 but not in en_US.UTF-8:

```
mfabian@taka:/tmp
$ gcc -lX11 compound.c -o compound
mfabian@taka:/tmp
$ LC_ALL=en_US.UTF-8 compound 
Xutf8TextListToTextProperty ret:1 value: 
XmbTextListToTextProperty ret:1 value: 
mfabian@taka:/tmp
$ LC_ALL=ja_JP.UTF-8 compound 
Xutf8TextListToTextProperty ret:0 value: 1b 24 28 42 22 65 
XmbTextListToTextProperty ret:0 value: 1b 24 28 42 22 65 
mfabian@taka:/tmp
$ 
```

**Comment by mike-fabian** [comment]
And then I found that copying the XLC_LOCALE file from the ja_JP.UTF-8 locale  to the en_US.UTF-8 locale makes it work:

```
mfabian@taka:/tmp
$ sudo cp /usr/share/X11/locale/en_US.UTF-8/XLC_LOCALE /usr/share/X11/locale/en_US.UTF-8/XLC_LOCALE.orig
mfabian@taka:/tmp
$ sudo cp /usr/share/X11/locale/ja_JP.UTF-8/XLC_LOCALE /usr/share/X11/locale/en_US.UTF-8/XLC_LOCALE
mfabian@taka:/tmp
$ LC_ALL=en_US.UTF-8 compound 
Xutf8TextListToTextProperty ret:0 value: 1b 24 28 42 22 65 
XmbTextListToTextProperty ret:0 value: 1b 24 28 42 22 65 
mfabian@taka:/tmp
$ 
```

I don’t yet know why and don’t know which unexpected side effects this could have.

**Comment by mike-fabian** [comment]
> > I also found that everything works fine when a CJK locale like ja_JP.UTF-8, ko_KR.UTF-8, ... is used but it does not work in en_US.UTF-8, de_DE.UTF-8, ...
> 
> That seems to work around it indeed.

This is probably the reason why I could never reproduce https://github.com/ibus/ibus/issues/1852 because I was most likely using ja_JP.UTF-8 locale when investigating it.

Setting only LC_CTYPE=ja_JP.UTF-8 is enough to make it work, the other LC_ variables and LANG are not needed.

**Comment by mike-fabian** [comment]
Only doing this change to the original `/usr/share/X11/locale/en_US.UTF-8/XLC_LOCALE` file makes it work already:

```diff
mfabian@taka:/usr/share/X11/locale/en_US.UTF-8
$ diff -u XLC_LOCALE.orig XLC_LOCALE
--- XLC_LOCALE.orig	2022-07-21 14:10:25.648477729 +0200
+++ XLC_LOCALE	2022-07-21 14:10:48.908260394 +0200
@@ -183,12 +183,12 @@
         ct_encoding     ISO8859-1:GR
 }
  
-#	cs2 class
-cs2     {
-        side            GR
-        length          1
-        ct_encoding     ISO8859-2:GR
-}
+##	cs2 class
+#cs2     {
+#        side            GR
+#        length          1
+#        ct_encoding     ISO8859-2:GR
+#}
  
 #	cs3 class
 cs3     {
mfabian@taka:/usr/share/X11/locale/en_US.UTF-8
$ 
```

**Comment by mike-fabian** [comment]
There seems to be nothing special with that `cs2 class` entry, I can uncomment other entries as well to make it work. For example only uncommenting like this also makes it work:

```diff
[mfabian@fedora en_US.UTF-8]$ diff -u XLC_LOCALE.orig XLC_LOCALE
--- XLC_LOCALE.orig	2022-07-21 13:10:14.867014873 +0200
+++ XLC_LOCALE	2022-07-21 14:56:30.497275703 +0200
@@ -211,13 +211,13 @@
         ct_encoding     ISO8859-5:GR
 }
  
-#	cs6 class
-cs6     {
-        side            GR
-        length          1
-        ct_encoding     KOI8-R:GR
-}
- 
+##	cs6 class
+#cs6     {
+#        side            GR
+#        length          1
+#        ct_encoding     KOI8-R:GR
+#}
+# 
 #	cs7 class
 cs7     {
         side            GR
[mfabian@fedora en_US.UTF-8]$ 
```

**Comment by mike-fabian** [comment]
The current state of the `/usr/share/X11/locale/en_US.UTF-8/XLC_LOCALE` file, which seems to cause that problem, is from this commit:

https://gitlab.freedesktop.org/xorg/lib/libx11/-/commit/418819558d2c60e58b4e3022ce0fadf2143488ac

And the original bug where this  came from is:

https://bugs.freedesktop.org/show_bug.cgi?id=2676
https://bugs.freedesktop.org/attachment.cgi?id=14467

**Comment by mike-fabian** [comment]
I do not understand what the problem with that old patch updating `/usr/share/X11/locale/en_US.UTF-8/XLC_LOCALE` is. It looks reasonable.

**Comment by mike-fabian** [comment]
New insights from Fujiwara San’s investigation:

> I think the reason is that KOI8-R does not have the escape sequence.
> https://gitlab.freedesktop.org/xorg/lib/libx11/-/blob/master/src/xlibi18n/lcCT.c#L58

That seems correct, there is no escape sequence for KOI8-R listed in lcCT.c and that seems to cause the problem.
Maybe no official escape sequence exists, I failed to find one. 

I think just removing that “cs6” section with the KOI8-R charset is OK.

It fixes the problem and I don’t think it will cause new problems. KOI8-R is a Russian charset. If this is removed, Russian will  fallback to the JISX0208.1983-0 charset:

```
#       cs16 class
cs12    {
        side            GR
        length          2
        ct_encoding     JISX0208.1983-0:GL; JISX0208.1983-0:GR;                         JISX0208.1983-1:GL; JISX0208.1983-1:GR
}
```

And this also supports Russian. So there should be no problems  with Cyrillic characters cause by that change.

**Comment by mike-fabian** [comment]
These locales have their own XLC_LOCALE file: 

`am_ET.UTF-8 cs_CZ.UTF-8 el_GR.UTF-8  fi_FI.UTF-8 km_KH.UTF-8 sr_RS.UTF-8`

but it is empty. According to locale.dir, they use the en_US.UTF-8 file:

```
[mfabian@fedora locale]$ for i in am_ET.UTF-8 cs_CZ.UTF-8 el_GR.UTF-8 fi_FI.UTF-8 km_KH.UTF-8 sr_RS.UTF-8; do grep ${i} locale.dir; done
en_US.UTF-8/XLC_LOCALE			am_ET.UTF-8
en_US.UTF-8/XLC_LOCALE:			am_ET.UTF-8
en_US.UTF-8/XLC_LOCALE			cs_CZ.UTF-8
en_US.UTF-8/XLC_LOCALE:			cs_CZ.UTF-8
en_US.UTF-8/XLC_LOCALE			el_GR.UTF-8
en_US.UTF-8/XLC_LOCALE:			el_GR.UTF-8
en_US.UTF-8/XLC_LOCALE			fi_FI.UTF-8
en_US.UTF-8/XLC_LOCALE:			fi_FI.UTF-8
en_US.UTF-8/XLC_LOCALE			km_KH.UTF-8
en_US.UTF-8/XLC_LOCALE:			km_KH.UTF-8
en_US.UTF-8/XLC_LOCALE			sr_RS.UTF-8
en_US.UTF-8/XLC_LOCALE			sr_RS.UTF-8@latin
en_US.UTF-8/XLC_LOCALE:			sr_RS.UTF-8
en_US.UTF-8/XLC_LOCALE:			sr_RS.UTF-8@latin
[mfabian@fedora locale]$ pwd
/usr/share/X11/locale
[mfabian@fedora locale]$ 
```

Fixing the `/usr/share/X11/locale/en_US.UTF-8/XLC_LOCALE` file fixes the problem for these locales as well.

**Comment by fujiwarat** [comment]
I added a link of the tentative patch to this issue.
It would be great if you could test the patch.

@mike-fabian Thank you very much for your evaluation and probably I think it's better to report the issue of Xutf8TextListToTextProperty() to libX11 upstream.

**Comment by ncfavier** [comment]
It works! Thanks.

**Comment by mike-fabian** [comment]
> It works! Thanks.

Did you try the patch to ibus or the removal of the KOI8-R charset from XLC_LOCALE?

**Comment by ncfavier** [comment]
I applied https://github.com/ibus/ibus/pull/2425 on top of 1.5.26.

**Comment by mike-fabian** [comment]
Merge request to fix the problem in en_US.UTF-8/XLC_LOCALE:

https://gitlab.freedesktop.org/xorg/lib/libx11/-/merge_requests/148

**Comment by fujiwarat** [comment]
> Merge request to fix the problem in en_US.UTF-8/XLC_LOCALE:

> https://gitlab.freedesktop.org/xorg/lib/libx11/-/merge_requests/148

Great. Thank you.

**Comment by mike-fabian** [comment]
> I applied #2425 on top of 1.5.26.

Good, I tried that too, worked for me as well.
